manticore/core/cpu/binja.py

```diff
@@ -380,6 +381,79 @@ def decode_instruction(self, pc):
         return insn
 
+    def execute(self):
+        '''
+        Decode, and execute one instruction pointed by register PC
+        '''
+        if issymbolic(self.PC):
+            raise ConcretizeRegister(self, 'PC', policy='ALL')
+
+        if not self.memory.access_ok(self.PC, 'x'):
+            raise InvalidMemoryAccess(self.PC, 'x')
+
+        self.publish('will_decode_instruction', self.PC)
+
+        insn = self.decode_instruction(self.PC)
+        self._last_pc = self.PC
+
+        self.publish('will_execute_instruction', insn)
+
+        # FIXME (theo) why just return here?
+        if insn.address != self.PC:
+            return
+
+        name = self.canonicalize_instruction_name(insn)
+        def fallback_to_emulate(*operands):
+            if (isinstance(self.disasm, BinjaILDisasm) and
+                    isinstance(insn, cs.CsInsn)):
+                # if we got a capstone instruction using BinjaILDisasm, it means
+                # this instruction is not implemented. Fallback to Capstone
+                self.FALLBACK(name, *operands)
+                # XXX after this point self.PC != self._last_pc but that is
+                # OK because we will update the PC  properly
+            else:
+                text_bytes = ' '.join('%02x'%x for x in insn.bytes)
+                logger.info("Unimplemented instruction: 0x%016x:\t%s\t%s\t%s",
+                            insn.address, text_bytes, insn.mnemonic, insn.op_str)
+
+                self.publish('will_emulate_instruction', insn)
+                self.emulate(insn)
+
+        implementation = getattr(self, name, fallback_to_emulate)
+
+        if logger.level == logging.DEBUG :
+            logger.debug(self.render_instruction(insn))
+            for l in self.render_registers():
+                register_logger.debug(l)
+
+        assert ((hasattr(insn, "sets_pc") and insn.sets_pc) or
+                self.PC == self._last_pc)
+
+        implementation(*insn.operands)
+
+        # In case we are executing IL instructions, we could iteratively
+        # invoke multiple instructions due to the tree form, thus we only
+        # want to increment the PC once, based on its previous position
+        # for CALLS and JUMPS the PC should have been set automatically
+        # so no need to do anything. Also, if there are pending instruction
+        if not isinstance(self.disasm, BinjaILDisasm):
+            return
+
+        # don't bump the PC if we are in an LLIL that has set it,
+        # or if there are pending IL insn in the queue. This is because
+        # for cases where we have other il instructions in the queue,
+        # such as when we get a divu insn, the PC + size will point
+        # to the next assembly instruction and not the next LLIL
+        #
+        # we might be executing a Capstone instruction at this point
+        # if we context-switched, so check the sets_pc attr
+        if not (hasattr(insn, "sets_pc") and
```

Comment:
There will never be a case where you will have an `il_queue` but not a `sets_pc`. Essentially
the only reason you check for `sets_pc` is because you could be executing a captstone instruction there, perhaps it makes more sense to check for that.